MDOT Form 2346 Address Page

maharielrosario opened the pull request:
## Description
This is the page where a veteran would select the address they want their hearing aid batteries and accessories to be shipped to. Features include a permanent address, temporary address, error messages for each field, and the ability to edit either address.

## Bugs
- Styling may or may not be accurate due to not having access to the specs of the design
- State between the radio buttons is not being shared

## Testing done
- Unit tests are passing

## Screenshots
![screenshot-localhost_3001-2020 03 24-20_09_09](https://user-images.githubusercontent.com/12755283/77488364-6071fc80-6e0b-11ea-84d5-b54634b844cd.png)
![screenshot-localhost_3001-2020 04 01-16_14_24](https://user-images.githubusercontent.com/12755283/78182320-f0c7c700-7433-11ea-9ff5-36dc33b7a8d5.png)



## Acceptance criteria
- Given that a veteran is ordering hearing aid batteries and accessories
- When they arrive at the address page
- Then they should be able to select a shipping address

## Definition of done
- [ ] There is a permanent address and temporary address component that displays address details received from the API
- [ ] There is a button to select a shipping address
- [ ] There is an edit link to edit the address fields
- [ ] There is a save button that saves any changes made to the address fields
- [ ] There is a cancel button that discards any changes made to the address fields

Commit:
code for Justin

Commit:
default state of order supplies page

Commit:
added conditional expand functionality

Commit:
adding ui descripton

Commit:
adding ui descripton

Commit:
updating JSON schema and form config

Commit:
updating JSON schema and form config

Commit:
adding stylying for the order supplies page

Commit:
added custom ui field

Commit:
removed title

Commit:
added all the ui elements needed for this scenaroio of the page

Commit:
removing supply fields from the reivew page

Commit:
removed reducer that is no long needed

Commit:
adding order accessories page

Commit:
changed schema and definitions

Commit:
comfirm page template and title change

Commit:
adding styles to pages

Commit:
created an accessoriesCustomFields directory

Commit:
adding checkbox button change styles

Commit:
changed checkbox checked color

Commit:
clean up existing schema and UI schema

Commit:
change name vetFields to schemaFields

Commit:
adding select array widget and changed schema to use it

Commit:
WIP part 1 address boxes

Commit:
WIP address boxes  part 2 and email UI

Commit:
WIP Address Cards Pt 3

Commit:
WIP address page pt 3.1

Commit:
fix date definition and rename
fields to addressFields and document error msg on address page

Commit:
WIP Addresss Page Pt 4

Commit:
radio btn state issue

Commit:
changed the action creator and removed veteranFullName

Commit:
adding proptypes to supplies

Commit:
implement custom reducer for radio btn state

Commit:
add military base info component

Commit:
modify email title and description

Commit:
add margin-bottom to military base dropdown

Commit:
Adding mapping to the order supplies page

Commit:
adding correct checkbox handler

Commit:
set up radio selections/
WIP for UI military base changes

Commit:
remove handleClick function

Commit:
fixed the checkbox issue

Commit:
adding array widget to accessories page

Commit:
adding separate widgets for supplies and accessories

Commit:
gating uiSchema updates with
the isMilitaryBaseChecked param

Commit:
remvoing string field

Commit:
adding conditonal logic

Commit:
adding conditonal logic for accessories page

Commit:
added another condition

rianfowler commented on src/applications/disability-benefits/2346/components/supplyCustomFields/deviceNameField.jsx:
```diff
@@ -0,0 +1,49 @@
+import React, { useState } from 'react';
+
+const deviceNameField = () => {
+  // eslint-disable-next-line react-hooks/rules-of-hooks
```
Why does this break the [rules of hooks](https://reactjs.org/docs/hooks-rules.html)?

mr0sari0 replied:
@rianfowler this actually came from the branch @mojaray2k had when I created this one. He could speak to this better.

U-DON replied:
The non-standard naming (not capitalizing component names) seems to be behind the linting error. https://github.com/department-of-veterans-affairs/vets-website/pull/11865#discussion_r399384510

Commit:
make updates to updateSchema and how it's called

Create gating functionality for each address component's ui schema.
Ensure updateSchema only runs when required props are available.
Swap out enumDisabled for ui:disabled

Commit:
bring in select batteries and select supplies pages

Refactored incoming code to meed standards

Commit:
Set up unit tests and still troubleshooting
military base UI bugs and state list changes

Commit:
comment out setData action for now

Commit:
Clean up unnecessary comments
and add class for disabled inputs

Commit:
update param listing for addressUISchema

Commit:
remove itemName param

Commit:
remove setData action

Commit:
make street and street 2 display on viewField

Commit:
move hideOnReview to inside of ui options

Commit:
prepping files to merge into address page branch

Commit:
WIP new address design pt 1

Commit:
schema enum  accepts country value instead of label

Commit:
clean up ui schemas and rename schema field

Commit:
show full country names instead of abbreviations

Commit:
WIP add missing files to address page

Commit:
show address info  on radio button labels

Commit:
troubleshooting radio btns
the top 2 are requiring country, state, and city
but they shouldn't as they are radios

bradl3yC committed:
Refactor names and exports

bradl3yC committed:
Rename folders appropriately

bradl3yC committed:
Refactor file names

bradl3yC committed:
Remove function unused function

bradl3yC committed:
refactor away from from2346reducer state usage

bradl3yC committed:
use callback for state and zip and add object fallback

bradl3yC committed:
Fix tests and remove eslint disable

bradl3yC committed:
Delete unused components

bradl3yC committed:
resolve merge conflicts

bradl3yC committed:
revert lint change

bradl3yC committed:
Add breadcrumbs

bradl3yC committed:
Use breadcrumb component and remove unused components

bradl3yC committed:
Style vet info page

bradl3yC commented:
@micahchiang @rianfowler @Mottie - This form is now working where we can progress all the way through.  There are still quite a few styling issues present and some content changes that need to happen.  I have managed to trim this down from 36 files to 24.  Can I get some reviews on this and will do the content / styling changes in follow up PRs?  @mr0sari0  for vis

kelsonic commented on src/applications/disability-benefits/2346/components/OrderAccessoriesPageContent.jsx:
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+
+const OrderAccessoriesPageContent = () => (
+  <div>
```
I think we can remove this extra wrapping `<div>` 🙂

kelsonic commented on src/applications/disability-benefits/2346/components/OrderSupplyPageContent.jsx:
```diff
@@ -1,6 +1,6 @@
+import React, { Component } from 'react';
 
-class orderSupplyPageContent extends React.Component {
+class OrderSupplyPageContent extends Component {
```
We could change this to a functional component if we want since there's no state being used 🙂

bradl3yC replied:
yassss

kelsonic commented on src/applications/disability-benefits/2346/components/PersonalInfoBox.jsx:
```diff
@@ -0,0 +1,76 @@
+  dateOfBirth,
+  ssnLastFour,
+}) => {
+  const nameString = [first, middle, last, suffix]
```
What do you think about changing this var name to `fullName`?

bradl3yC replied:
works for me

kelsonic commented on src/applications/disability-benefits/2346/components/PersonalInfoBox.jsx:
```diff
@@ -0,0 +1,76 @@
+  middle: '',
+  gender: '',
+  dateOfBirth: '',
+  ssnLastFour: '1234',
```
Do we want this as a default prop? Just confirming 🙂

bradl3yC replied:
@kelsonic yeah I used that because its not included in the betamocks and i needed it for styling.  I can remove.

kelsonic commented on src/applications/disability-benefits/2346/config/form.js:
```diff
@@ -34,10 +41,12 @@ const formChapters = {
 const formPages = {
```
What do you think about `formPageTitlesLookup` or something similar? I think adding `title` to this variable name may make it clearer 🙂

kelsonic replied:
After reading more of the code, I'm wondering if we could have something like:
```js
const formPagesLookup = {
  addBatteriesPage: { path: '/some/path', title: 'Add batteries to your order' },
  addAccessoriesPage: { path: '/some/path', title: 'Add accessories to your order' },
  // etc.
}
```

kelsonic commented on src/applications/disability-benefits/2346/config/form.js:
```diff
@@ -1,30 +1,37 @@
+  selectedAddressUI,
+} = UIDefinitions.sharedUISchemas;
 
 const formChapters = {
```
https://github.com/department-of-veterans-affairs/vets-website/pull/11999/files#r406809739
same here ;)

bradl3yC committed:
Fix for comments

kelsonic commented on src/applications/disability-benefits/2346/constants/index.js:
```diff
@@ -6,71 +11,72 @@ export const FETCH_REORDER_BATTERY_AND_ACCESSORIES_INFORMATION =
-  'Armed Forces Europe (AE)',
-  'Armed Forces Pacific (AP)',
-];
+export const militaryStates = states.USA.filter(
```
So we actually have a PR that's trying to consolidate all of the state fields across the VA! https://github.com/department-of-veterans-affairs/vets-website/pull/12178

I'll tag this PR in that PR! 🙂

bradl3yC replied:
@kelsonic yeah! I think that work will make some changes to this file once merged

Mottie replied:
Actually I added a comment in that PR that I specifically didn't modify this form's folder to prevent merge conflicts since I know you're actively developing in it. I don't like dealing with merge conflicts :D

bradl3yC replied:
@Mottie 💯

kelsonic commented on src/applications/disability-benefits/2346/containers/App.jsx:
```diff
@@ -1,12 +1,27 @@
-    </RoutedSavableApp>
+    <>
+      <Breadcrumbs>
+        {[
```
You can actually remove the wrapping `{[ ]}` and remove the key properties on all of the children! 🙂

kelsonic replied:
Also we'll need `hrefs` here as well as `rel="noreferrer noopener"` for links that are not va.gov based 📈

kelsonic commented on src/applications/disability-benefits/2346/containers/ConfirmationPage.jsx:
```diff
@@ -19,7 +18,6 @@ export class ConfirmationPage extends React.Component {
   componentDidMount() {
     focusElement('.schemaform-title > h1');
     scrollToTop();
```
You can use our reusable module for this:
```js
import scrollToTop from 'platform/utilities/ui/scrollToTop';
```

bradl3yC replied:
Good catch, I actually need to remove this

kelsonic commented on src/applications/disability-benefits/2346/containers/ConfirmationPage.jsx:
```diff
@@ -19,7 +18,6 @@ export class ConfirmationPage extends React.Component {
   componentDidMount() {
     focusElement('.schemaform-title > h1');
```
You can use our reusable module for this:
```js
import { focusElement } from 'platform/utilities/ui';
```

I think we may be missing an import for this function 🙂

kelsonic commented on src/applications/disability-benefits/2346/helpers/index.js:
```diff
@@ -71,3 +72,20 @@ export function validateZIP(errors, zip) {
 }
+
+export const newAddressHider = formData => {
+  const address = get('newAddress', formData) || {};
```
Instead of using our `get` module you could do `formData?.newAddress` if you'd like! 🙂

bradl3yC replied:
Its also weird that there are 2 imports from utlities/data - Ill fix that too

bradl3yC committed:
refactor confirmation page

kelsonic commented on src/applications/disability-benefits/2346/helpers/index.js:
```diff
@@ -71,3 +72,20 @@ export function validateZIP(errors, zip) {
+  // return true if living at view:militaryBaseInfo object
+  return Object.keys(address).every(entry => {
+    const value = address[entry] || '';
+    return typeof value === 'string' ? value === '' : true;
```
`return typeof value !== 'string'`, although this function may need some love to ensure it's doing what's desired! 🙂

bradl3yC replied:
Yeah I have no idea whats going on here - I will have to dig into this further

bradl3yC committed:
refactor helper

bradl3yC committed:
Fix breadcrumbs

kelsonic commented on src/applications/disability-benefits/2346/helpers/index.js:
```diff
@@ -71,3 +72,20 @@ export function validateZIP(errors, zip) {
+  });
+};
+
+export const updateRadioLabels = (formData, name) => {
```
May want to explore omitting/giving love to this function, too 🙂

bradl3yC replied:
💯 function love incoming

kelsonic commented on src/applications/disability-benefits/2346/sass/form-2346.scss:
```diff
@@ -8,14 +8,14 @@
 
 .order-background {
-  background-color: $color-gray-lightest;
+  background-color: #f1f1f1;
```
If we can't use design utilities, we may want to indicate this to the design team to change their spec

kelsonic commented on src/applications/disability-benefits/2346/sass/form-2346.scss:
```diff
@@ -39,8 +39,8 @@ label {
-  background: $color-white !important;
-  color: $color-primary !important;
+  background: #fff !important;
+  color: #0071bb !important;
```
https://github.com/department-of-veterans-affairs/vets-website/pull/11999/files#r406826151

kelsonic commented on src/applications/disability-benefits/2346/sass/form-2346.scss:
```diff
@@ -53,12 +53,12 @@ label {
 
 .required {
-  color: $color-secondary-dark;
+  color: red;
```
https://github.com/department-of-veterans-affairs/vets-website/pull/11999/files#r406826151

kelsonic commented on src/applications/disability-benefits/2346/sass/form-2346.scss:
```diff
@@ -53,12 +53,12 @@ label {
 
 [type=checkbox]:checked + label::before, [type=radio]:checked + label::before {
-   background-color: $color-white;
+   background-color: #fff;
```
https://github.com/department-of-veterans-affairs/vets-website/pull/11999/files#r406826151

kelsonic commented on src/applications/disability-benefits/2346/sass/form-2346.scss:
```diff
@@ -97,6 +103,10 @@ label {
     }
 
+    input[type="text"]:disabled {
+      background-color: #f1f1f1;
```
https://github.com/department-of-veterans-affairs/vets-website/pull/11999/files#r406826151

kelsonic commented on src/applications/disability-benefits/2346/schemas/2346-schema.json:
```diff
@@ -229,68 +250,131 @@
           "type": "string",
-          "enum": ["USA", "CAN", "MEX"]
+          "default": "USA",
+          "enum": [
```
https://github.com/department-of-veterans-affairs/vets-website/pull/11999/files#r406813416

kelsonic commented on src/applications/disability-benefits/2346/schemas/address-schema.js:
```diff
@@ -0,0 +1,258 @@
+ * 4. addressSchema - data model for address schema.
+ */
+
+const MILITARY_STATES = Object.entries(ADDRESS_DATA.states).reduce(
```
https://github.com/department-of-veterans-affairs/vets-website/pull/11999/files#r406813416

kelsonic commented on src/applications/disability-benefits/2346/schemas/address-schema.js:
```diff
@@ -0,0 +1,258 @@
+};
+
+const MilitaryBaseInfo = () => (
+  <div className="vads-u-padding-x--2p5">
```
We could omit this wrapping div and add a className to `AdditionalInfo` below if you want! 🙂

kelsonic commented on src/applications/disability-benefits/2346/schemas/address-schema.js:
```diff
@@ -0,0 +1,258 @@
+      status="info"
+      triggerText="Learn more about military base addresses"
+    >
+      <span>
```
Do we want to wrap this text with `span` or should we use `p` here?

kelsonic commented on src/applications/disability-benefits/2346/schemas/address-schema.js:
```diff
@@ -0,0 +1,258 @@
+    },
+    country: {
+      type: 'string',
+      enum: countries.map(country => country.value),
```
nice!

kelsonic commented on src/applications/disability-benefits/2346/schemas/address-schema.js:
```diff
@@ -0,0 +1,258 @@
+ * @param {boolean} isMilitaryBaseAddress represents whether or not the form page requires the address to support the option of military address.
+ */
+
+const MILITARY_BASE_PATH = '[view:livesOnMilitaryBase]';
```
Hmm, did we mean to make this a stringified array here? 🙂

bradl3yC replied:
Maybe? 😂 Ill find out

kelsonic commented on src/applications/disability-benefits/2346/schemas/address-schema.js:
```diff
@@ -0,0 +1,258 @@
+) => {
+  // As mentioned above, there are certain fields that depend on the values of other fields when using updateSchema, replaceSchema, and hideIf.
+  // The two constants below are paths used to retrieve the values in those other fields.
+  const livesOnMilitaryBasePath = `${path}${MILITARY_BASE_PATH}`;
```
Ah, I see... I would assume that this variable is a boolean based off the naming... also we seem to be trying to grab a `formData` property from this string farther down, a little confusing https://github.com/department-of-veterans-affairs/vets-website/pull/11999/files#diff-357ddb3270f3e834209fd32d3cdbb1cfR165

kelsonic commented on src/applications/disability-benefits/2346/schemas/definitions/2346UI.js:
```diff
@@ -0,0 +1,130 @@
+const { viewAddAccessoriesField, viewAddBatteriesField } = schemaFields;
+
+const emailUITitle = (
+  <>
```
We don't need this wrapping `<>` tag 🙂

bradl3yC replied:
nobody likes an empty tag anyway

kelsonic commented on vets-website.code-workspace:
```diff
@@ -0,0 +1,7 @@
+			"path": "."
+		}
+	]
+}
```
Did we mean to commit this file? Looks like it's vscode-specific 🙂

bradl3yC replied:
Revert inc - this existed before I jumped in, good catch.

kelsonic commented:
I left a few comments! Do you have time to jump on a call to walk me through this PR or could you make a Quicktime screen video of this PR in action? @bradl3yC @mr0sari0 Would love to see this PR's resultant functionality 🙂

bradl3yC commented:
@kelsonic yeah! Let me tidy this up based on your comments and we can jump on a call.

bradl3yC committed:
Refactor helper and remove vscode file

bradl3yC committed:
Refactor updateRadioLabels helper func

bradl3yC committed:
correct css utility usage

bradl3yC committed:
correct css utility usage

bradl3yC committed:
remove unnecessary div

bradl3yC committed:
remove empty tag

kelsonic commented:
LGTM! Thanks for making all the updates! I've also heard there will be a follow-up PR with even more updates, exciting! Nice work 🎉
